r/R/dplyr-count-tally.R

```diff
@@ -0,0 +1,70 @@
+# Licensed to the Apache Software Foundation (ASF) under one
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+
+# The following S3 methods are registered on load if dplyr is present
+
+count.arrow_dplyr_query <- function(x, ..., wt = NULL, sort = FALSE, name = NULL) {
+  if (!missing(...)) {
+    out <- group_by(x, ..., .add = TRUE)
+  } else {
+    out <- x
+  }
+
+  out <- tally(out, wt = {{ wt }}, sort = sort, name = name)
+
+  # Restore original group vars
+  gv <- dplyr::group_vars(x)
+  if (length(gv)) {
+    out$group_by_vars <- gv
+  }
+
+  out
+}
+
+count.Dataset <- count.ArrowTabular <- count.arrow_dplyr_query
+
+tally.arrow_dplyr_query <- function(x, wt = NULL, sort = FALSE, name = NULL) {
+  n <- tally_n(x, {{ wt }})
+  name <- check_name(name, group_vars(x))
+
+  out <- local({
+    old.options <- options(dplyr.summarise.inform = FALSE)
+    on.exit(options(old.options))
+    summarise(x, !!name := !!n)
```

Comment:
This needs to call dplyr:: since it's Suggests and we need that to get the dispatch to work.

```suggestion
    dplyr::summarise(x, !!name := !!n)
```

Also, what I meant about inlining before: I meant more like this, which would remove one level of expression boxing and unboxing:

```
wt <- enquo(wt)
if (quo_is_null(wt)) {
  summarize(x, !!name = n())
} else {
  summarize(x, !!name = sum(!!wt, na.rm = TRUE))
}
```

Reply:
What's inlining?

Reply:
Oh, adding the code "in line" rather than factored out into a function.

Reply:
Right, I see.  I tried implementing it like this:

```
tally.arrow_dplyr_query <- function(x, wt = NULL, sort = FALSE, name = NULL) {
  
  wt <- enquo(wt)

  name <- check_name(name, group_vars(x))

  if (quo_is_null(wt)) {
    out <- dplyr::summarize(x, {{name}} := n())
  } else {
    out <- dplyr::summarize(x, {{name}} := sum(!!wt, na.rm = TRUE))
  }

  if (sort) {
    arrange(out, desc(!!sym(name)))
  } else {
    out
  }
}
```

But keep getting a weird error `No arrow R6 class named 'p%'` which apparently comes from `r/src/arrow_cpp11.h` which has absolutely baffled me.  Any idea where I should be looking next with this?

Reply:
Doesn't ring a bell. Though if I read this correctly, it should be `!!name` instead of `{{name}}`, maybe that's related. If you're still having trouble I can pull and take a look.